File: PythonDesignPatterns/chainOfResponsibility.py


from abc import ABC, abstractmethod

class Handler(ABC):

    def __init__(self):
        self.next_handler=None

    def set_next(self,handler):
        self.next_handler=handler
        return handler
    @abstractmethod
    def handle(self):
        pass


class Authentication_check(Handler):

    def handle(self,request):
        if request["authenticated"]==True:
            print("Authentication Passed")
            self.next_handler.handle(request)
        else:
            print("Authentication Failed")


class Role_check(Handler):

    def handle(self,request):
        if request["role"]=="admin":
            print("Access granted")
            self.next_handler.handle(request)
        else:
            print("Access denied")




class RateLimit_check(Handler):

    def handle(self,request):
        if request["requests_made"]<True:
            print("Rate Limit passed")
            self.next_handler.handle(request)
        else:
            print("Rate Limit Exceeded")



request = {"authenticated": True, "role": "admin", "requests_made": 3}


auth=Authentication_check()
auth.set_next(Role_check()).set_next(RateLimit_check())

auth.handle(request)

File: NodeDesignPatterns/chainOfResponsibility.ts

type AppRequest = {
    authenticated: boolean;
    role: string;
    requests_made: number;
};

abstract class Handler{
    next_handler: Handler| null=null;

    set_next(handler:Handler):Handler{
        this.next_handler=handler
        return handler

    }

    abstract handle(request:AppRequest):void
}


class Authentication extends Handler{
    handle(request:AppRequest):void{
        if(request.authenticated==true){
            console.log("Authentication passed")
            if (this.next_handler){
                this.next_handler.handle(request)
            }
            else{
                console.log("end of the chain")
            }
            
        }
        else{
            console.log("Authentication failed")
        }

    }
}





class Role extends Handler{
    handle(request:AppRequest):void{
        if(request.role==="admin"){
            console.log("Role Authienticated")
            if (this.next_handler){
                this.next_handler.handle(request)
            }
            else{
                console.log("end of the chain")
            }
            
        }
        else{
            console.log("Access denied")
        }

    }
}






class RateLimit extends Handler{
    handle(request:AppRequest):void{
        if(request.requests_made<10){
            console.log("Rate Limit passed")
            if (this.next_handler){
                this.next_handler.handle(request)
            }
            else{
                console.log("end of the chain")
            }
        }
        else{
            console.log("Rate Limit Exeeded")
        }

    }
}



let auth= new Authentication()
auth.set_next(new Role()).set_next(new RateLimit())
const request: AppRequest = {
    authenticated: true,
    role: "admin",
    requests_made: 3
};

auth.handle(request)